taiga2/controllers/models_controller.py

```diff
@@ -736,48 +760,140 @@ def format_renamed(datafiles):
+
+def format_datafile_id(
+    dataset_permaname: str,
+    dataset_version: DatasetVersion,
+    datafile_name: Optional[str],
+):
+    name_parts = {
+        "dataset_permaname": dataset_permaname,
+        "dataset_version": dataset_version,
+        "datafile_name": datafile_name,
+    }
+
+    id_format = (
+        DATAFILE_ID_FORMAT
+        if datafile_name is not None
+        else DATAFILE_ID_FORMAT_MISSING_DATAFILE
+    )
+
+    return id_format.format(**name_parts)
+
+
+def get_previous_version_upload_datafiles(
+    dataset_version_datafiles: List[DataFile],
+    dataset_permaname: str,
+    dataset_version: str,
+) -> List[UploadVirtualDataFile]:
+    previous_version_taiga_ids = [
+        {
+            "taiga_id": format_datafile_id(
+                dataset_permaname, dataset_version, datafile.name
+            )
+        }
+        for datafile in dataset_version_datafiles
+    ]
+
+    previous_version_datafiles = (
+        [UploadVirtualDataFile(f) for f in previous_version_taiga_ids]
+        if previous_version_taiga_ids is not None
+        else None
     )
 
-    comments = _format_datafile_diff(datafile_diff)
+    return previous_version_datafiles
 
-    new_dataset_version = add_dataset_version(
-        dataset_id=dataset_id,
-        datafiles_ids=all_datafile_ids,
-        new_description=new_description,
-        changes_description=changes_description,
+
+def generate_previous_version_upload_files(existing_taiga_id, sid):
+    try:
+        data_file = get_datafile_by_taiga_id(existing_taiga_id, one_or_none=True)
+    except InvalidTaigaIdFormat as ex:
+        api_error(
+            "The following was not formatted like a valid taiga ID: {}".format(
+                ex.taiga_id
+            )
+        )
+
+    if data_file is None:
+        api_error("Unknown taiga ID: " + existing_taiga_id)
+
+    return add_upload_session_virtual_file(
+        session_id=sid, filename=data_file.name, data_file_id=data_file.id
     )
 
-    activity = VersionAdditionActivity(
-        user_id=current_user.id,
-        dataset_id=new_dataset_version.dataset_id,
-        type=Activity.ActivityType.added_version,
-        dataset_description=new_description
-        if new_description != latest_dataset_version.description
-        else None,
-        dataset_version=new_dataset_version.version,
-        comments=comments,
+
+def get_previous_version_and_added_datafiles(
+    dataset_version: DatasetVersion,
+    dataset_id: str,
+    datafile_names: List[str],
+    session_id: str,
+) -> List[DataFile]:
+    version = None
+    if dataset_version is None:
+        version = get_latest_dataset_version(dataset_id)
+    else:
+        assert dataset_version.version is not None
+        version = get_dataset_version_by_dataset_id_and_version(
+            dataset_id, dataset_version.version
+        )
+
+    previous_version_datafiles = get_previous_version_upload_datafiles(
```

Comment:
This appears to return instances of `UploadVirtualDataFile` however, this is an external structure, intended for the API. Since this code is all internal to Taiga, you can eliminate a lot of code by returning the DataFiles instead and using those to add to the session instead of translating datafiles -> taiga_id -> datafile.

I think that will be simpler and make this much easier to follow what's going on. I have a hunch that you might have ended up with this as a result of trying to lift logic from taigapy into this class. However, I think it can all be done much simpler. 

Happy to talk through this more if that'd be helpful.